Dockerfile

```diff
@@ -10,31 +10,46 @@ RUN apt-get update \
         libssl-dev \
         git \
     && apt-get clean \
-    && rm -rf /var/lib/apt/lists/*
+    && rm -rf /var/lib/apt/lists/* \
+    && python -m venv /opt/venv
 
-COPY requirements.txt /opt/CTFd/
-
-RUN pip install -r requirements.txt --no-cache-dir
+ENV PATH="/opt/venv/bin:$PATH"
 
 COPY . /opt/CTFd
 
-# hadolint ignore=SC2086
-RUN for d in CTFd/plugins/*; do \
+RUN pip install --no-cache-dir -r requirements.txt \
+    && for d in CTFd/plugins/*; do \
         if [ -f "$d/requirements.txt" ]; then \
-            pip install -r $d/requirements.txt --no-cache-dir; \
+            pip install --no-cache-dir -r "$d/requirements.txt";\
         fi; \
     done;
 
-# hadolint ignore=DL3059
-RUN adduser \
-    --disabled-login \
-    -u 1001 \
-    --gecos "" \
+
+FROM python:3.9-slim-buster as release
+WORKDIR /opt/CTFd
+
+# hadolint ignore=DL3008
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+        libffi6 \
+        libssl1.1 \
```

Comment:
Do we specifically need to pin versions here or could we just let the package manager pick the latest?

Reply:
Interesting question.
On apt repository of Debian latest tag only have for Foreign Function Interface library (development files).
I not think justification for this choice from apt repository. On Ubuntu, it's the same.

For libffi example link
https://packages.debian.org/source/buster/libffi
Runtime package is named libffi6 and dev package is libffi-dev.

For libssl exatly same.
https://packages.debian.org/buster/libssl-dev
https://packages.debian.org/buster/libssl1.1

Extra:
On apk repository of Alpine have latest tags for dev and runtime packages.

Reply:
But for check what packages have on apt reposity use
```
apt-cache search libffi
apt-cache search libssl
```

For libffi example
```
$ apt-cache search libffi
libffi-dev - Foreign Function Interface library (development files)
libffi6 - Foreign Function Interface library runtime
libffi6-dbg - Foreign Function Interface library runtime (debug symbols)
ffindex-dbg - simple index/database for huge amounts of small files (debug)
libbfio-dbg - Library to provide basic input/output abstraction (debug)
libffindex0 - library for simple index/database for huge amounts of small files
libffindex0-dev - library for simple index/database for huge amounts of small files (development)
libjffi-java - Java Foreign Function Interface
libjffi-jni - Java Foreign Function Interface (JNI library)
```

Reply:
I think this makes sense. I just need to spend some time to test this out because there may be some edge case that we are missing.

Reply:
I have tested this out and it seems to be working properly.